File: data/src/main/java/fiveagency/internship/food/data/database/dao/MovieDao.java
package fiveagency.internship.food.data.database.dao;

import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;

import java.util.List;

import fiveagency.internship.food.data.database.model.DbMovie;
import io.reactivex.Single;

@Dao
public interface MovieDao {

    @Query("SELECT * FROM movie")
    Single<List<DbMovie>> getAllMovies();

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    void insertAllMovies(DbMovie... dbMovie);

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    void insertAllMovies(List<DbMovie> dbMovies);

    @Delete
    void deleteMovie(DbMovie dbMovie);

    @Query("UPDATE movie SET personalNote = :personalNote WHERE id = :movieId")
    void setPersonalNote(String personalNote, int movieId);

    @Query("SELECT * FROM movie WHERE id = :movieId")
    Single<DbMovie> getMovie(int movieId);

    @Query("SELECT * FROM movie WHERE id = :movieId")
    Single<List<DbMovie>> movieExists(int movieId);
}

File: data/src/main/java/fiveagency/internship/food/data/database/dao/MovieDao.kt
package fiveagency.internship.food.data.database.dao

import com.google.firebase.firestore.FirebaseFirestore
import de.aaronoe.rxfirestore.getFlowable
import fiveagency.internship.food.data.database.mappers.MovieModelMapper
import fiveagency.internship.food.data.database.model.DbMovie
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.Single

private const val MOVIE_COLLECTION_NAME = "movies"

class MovieDao(private val firebaseFirestore: FirebaseFirestore, private val movieModelMapper: MovieModelMapper) {

    fun insertAllMovies(movies: List<DbMovie>) {
        movies.forEach {
            firebaseFirestore.collection(MOVIE_COLLECTION_NAME)
                .document(it.id.toString())
                .set(it)
        }

    }

    fun deleteMovie(dbMovie: DbMovie) {
        firebaseFirestore.collection(MOVIE_COLLECTION_NAME)
            .document(dbMovie.id.toString())
            .delete()
    }

    fun setPersonalNote(personalNote: String, movieId: Int) {
        firebaseFirestore
            .collection(MOVIE_COLLECTION_NAME)
            .document(movieId.toString())
            .get()
            .addOnSuccessListener {
                val dbMovie: DbMovie = it.toObject(DbMovie::class.java) ?: DbMovie.EMPTY
                dbMovie.personalNote = personalNote
                insertMovie(dbMovie)
            }
    }

    private fun insertMovie(dbMovie: DbMovie) {
        firebaseFirestore
            .collection(MOVIE_COLLECTION_NAME)
            .document(dbMovie.id.toString())
            .set(dbMovie)
    }

    fun getMovie(movieId: Int): Single<DbMovie> {
        return Single.create { singleEmitter ->
            firebaseFirestore
                .collection(MOVIE_COLLECTION_NAME)
                .document(movieId.toString())
                .get()
                .addOnSuccessListener {
                    val dbMovie: DbMovie = it.toObject(DbMovie::class.java) ?: DbMovie.EMPTY
                    singleEmitter.onSuccess(dbMovie)
                }
                .addOnFailureListener {
                    singleEmitter.onError(it)
                }
        }
    }

    fun getAllMovies(): Flowable<List<DbMovie>> {
        return firebaseFirestore.collection(MOVIE_COLLECTION_NAME).getFlowable(BackpressureStrategy.LATEST)
    }
}